src/TorchSharp/DisposeScope.cs

```diff
@@ -0,0 +1,301 @@
+﻿// Copyright (c) .NET Foundation and Contributors.  All Rights Reserved.  See LICENSE in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using TorchSharp.Utils;
+
+namespace TorchSharp
+{
+    /// <summary>
+    /// Keeps track of all disposables that are in the current scope - the dispose scopes can be nested and the
+    /// nesting functionality is mainly managed by DisposeScopeManager.
+    /// </summary>
+    public class DisposeScope : IDisposable
+    {
+        private readonly DisposeScopeManager _disposeScopeManager;
+
+        public DisposeScope(DisposeScopeManager disposeScopeManager)
+        {
+            _disposeScopeManager = disposeScopeManager;
+            if (disposeScopeManager.DisposeScopeStack.Count > 0) {
+                OuterScope = disposeScopeManager.DisposeScopeStack.Peek();
+            }
+        }
+
+        /// <summary>
+        /// The outer scope with relation to this scope.
+        /// </summary>
+        internal DisposeScope OuterScope { get; set; }
+
+        /// <summary>
+        /// The disposables that are scheduled for disposing.
+        /// </summary>
+        /// TODO: There is a ReferenceEqualityComparer coming in .NET 6, use that!
+        internal HashSet<IDisposable> Disposables { get; private set; } =
```

Comment:
NIT: I don't think you need `private set;` if you ever only assign the property in the initializer.